insights/combiners/tmpfilesd.py

```diff
@@ -0,0 +1,74 @@
+"""
+Tmpfilesd configuration
+=======================
+
+The tmpfilesd files are normally available to rules as a list of TmpFilesd
+objects. This combiner turns those into one set of data, and provides a
+``find_file()`` method to search for that filename among all the files.
+"""
+
+from insights.core.plugins import combiner
+from insights.parsers.tmpfilesd import TmpFilesD
+
+
+@combiner(TmpFilesD)
+class AllTmpFiles(object):
+    """
+    Combiner for accessing all the tmpfilesd configuration files. Configuration
+    files can be found in three directories: ``/usr/lib/tmpfiles.d``,
+    ``/run/tmpfiles.d``, and ``/etc/tmpfiles.d``. Each directory overrides the
+    settings in the previous directory. This combiner checks for and accounts
+    for this behavior.
+
+    Attributes:
+        files(set): the set of files found in all data files.
+        rules(dict): a dictionary of rules using the config file as the key
+        file_paths(list): a list of the file paths for the configurations.
+    """
+    def __init__(self, tmpfiles):
+        rules = {}
+        files = set([])
+        self.tmpfiles = tmpfiles
+        file_paths = []
+        # Files must be sorted so that /etc/tmpfiles.d overrides /run/tmpfiles.d
+        # which overides /usr/lib/tmpfiles.d. To do this, we have to list file
+        # paths in reverse order.
+        for tmpfile in sorted(tmpfiles, key=lambda f: f.file_path, reverse=True):
+            file_paths.append(tmpfile.file_path)
+            rules[tmpfile.file_path] = []
+            for new_rule in tmpfile.rules:
+                _file = (new_rule['path'])
+                if _file in files:
+                    [rules[tmpfile.file_path].remove(_rule) for _rule in rules[tmpfile.file_path] if _rule['path'] == _file]
```

Comment:
I don't think this `.remove` will work as expected according to current structure (`file_path` as the key), since at this time in each pass of loop the `rules[tmpfile.file_path]` is still an empty `[]`.

As per current structure, you have to iterate all the data which is already stored into `rules` to remove the duplicate rules.

Would you mind to try to add a new key `file_path` for each `rule`? It's redundant, but could simplify the logic a little. You can take a look at the `combiners/httpd_conf.py` for a reference.

Reply:
@SteveHNH - Sorry for my poor description. The `.remove` might still do not work as what you want, if you want to override the rule with the same `file` but defined in different `file_path` (it's not because of the empty `[]`, sorry)

The `tmpfile.file_path` is an iteration of the `outer` loop, it will be always the same one for the `inner` loop. So, it only works for if there are re-defined rules inner the same file, but does not work for different files.

For example, in case of the test `all_data = AllTmpFiles([data1, data3])`. It cannot remove the rule `r! /tmp/.sap*`.

It's OK to me if you just want to remove the duplicated rules inner the same file. But you might need to note that it won't remove between different file.